### PR TITLE
Use set to clearify states field semantic

### PR DESCRIPTION
**Pull request by JelF**
- fixed code to match specs
- used set for Statum::StateDefiner#events
- improved documentation

### CODE REVIEW

**Comment by JelF** [comment]
Should i remove `Statum::Event.before?` or cover it with test?

File: lib/statum/machine.rb [hunk 1]
```diff
@@ -45,9 +45,18 @@ def fire!(instance, name)
...
-      instance.instance_eval(&event.after) if event.after?
+      exec_hook(instance, event.after)
+    end
+
+    def exec_hook(instance, hook)
+      if hook.nil?
+      elsif hook.arity.zero?
+        instance.instance_exec(&hook)
+      else
+        instance.instance_exec(instance, &hook)
```
Comment (nulldef) [inline lib/statum/machine.rb:61]: Please, note this feature (passed argument) in README

**Review by nulldef — COMMENTED** [review]
@JelF thanks for your PR! I think we don't need more `before?` and `after?` methods and you can remove them from `Statum::Event`. Also note the comment I've left in code. 
After then we can merge this request, I suggest.

**Comment by JelF** [comment]
There is also a question, should non-zero arity procs be executed in `instance_exec` context? I think, there is no reason to pass instance both as receiver and as argument

**Comment by JelF** [comment]
> There is also a question, should non-zero arity procs be executed in instance_exec context? I think, there is no reason to pass instance both as receiver and as argument

Sure beter to convert this question into an issue

**Comment by nulldef** [comment]
Yes, I think we should pass instance into hooks always. I've already left a comment about it in #8.
After merging this PR I can test it in multiple ruby versions.

PS Which ruby version did you use? In some versions (ex 2.3.1) I had troubles with `instance_exec`

**Comment by JelF** [comment]
> PS Which ruby version did you use? In some versions (ex 2.3.1) I had troubles with instance_exec

2.4

I can update travis.yml to check 2.3 and 2.2

**Comment by nulldef** [comment]
@JelF no, thanks, I have to resolve it by self. Can you resolve #8 in this PR?

**Comment by nulldef** [comment]
Expired

**Pull request closed without merging**